packages/router/src/router.tsx

```diff
@@ -42,57 +137,119 @@ interface PrivateProps {
+    >
+      {children}
+    </PrivateContextProvider>
+  )
 }
 
-interface PrivatePageLoaderProps {
-  useAuth: typeof useAuthHook
-  unauthenticatedRoute: (args?: Record<string, string>) => string
-  role: string | string[]
-  whileLoading?: () => ReactElement | null
+function isRoute(
+  node: React.ReactNode
+): node is React.ReactElement<RouteProps> {
+  return isReactElement(node) && node.type === Route
 }
 
-const PrivatePageLoader: React.FC<PrivatePageLoaderProps> = ({
+interface RouterProps extends RouterState {}
+
+const Router: React.FC<RouterProps> = ({
   useAuth,
-  unauthenticatedRoute,
-  role,
-  whileLoading = () => null,
+  paramTypes,
+  pageLoadingDelay,
   children,
 }) => {
-  const { loading, isAuthenticated, hasRole } = useAuth()
+  const flatChildArray = flattenAll(children)
+  const shouldShowSplash =
+    flatChildArray.length === 1 &&
+    isRoute(flatChildArray[0]) &&
+    flatChildArray[0].props.notfound
 
-  if (loading) {
-    return whileLoading()
+  if (shouldShowSplash) {
+    return <SplashPage />
   }
 
-  if (
-    (isAuthenticated && !role) ||
-    (isAuthenticated && role && hasRole(role))
-  ) {
-    return <>{children || null}</>
-  }
+  flatChildArray.forEach((child) => {
+    if (isRoute(child)) {
+      const { name, path } = child.props
+
+      if (path) {
+        // Check for issues with the path.
+        validatePath(path)
+
+        if (name && path) {
+          namedRoutes[name] = (args = {}) => replaceParams(path, args)
+        }
+      }
+    }
+  })
 
-  const currentLocation =
-    window.location.pathname + encodeURIComponent(window.location.search)
   return (
-    <Redirect to={`${unauthenticatedRoute()}?redirectTo=${currentLocation}`} />
+    <RouterContextProvider
+      useAuth={useAuth}
+      paramTypes={paramTypes}
+      pageLoadingDelay={pageLoadingDelay}
+    >
+      <LocationProvider>
+        <NotFoundChecker>
+          <>{children}</>
+        </NotFoundChecker>
+      </LocationProvider>
+    </RouterContextProvider>
   )
 }
 
-interface RouterProps {
-  useAuth?: typeof useAuthHook
-  paramTypes?: Record<string, ParamType>
-  pageLoadingDelay?: number
-}
+const NotFoundChecker: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const location = useLocation()
+  const routerState = useRouterState()
+
+  let foundMatchingRoute = false
+  let NotFoundPage: PageType | undefined = undefined
+  const flatChildArray = flattenAll(children)
+
+  flatChildArray.forEach((child) => {
+    if (isRoute(child)) {
+      const { path } = child.props
+
+      if (path) {
+        const { match } = matchPath(
+          path,
+          location.pathname,
+          routerState.paramTypes
+        )
 
-const Router: React.FC<RouterProps> = (props) => (
-  <Location>
-    {(locationContext: LocationContextType) => (
-      <RouterImpl {...locationContext} {...props} />
-    )}
-  </Location>
-)
+        if (match) {
+          foundMatchingRoute = true
+        }
```

Comment:
At this point we don't need to loop over the children anymore, since we've found the route?

Reply:
Sure. Oldschool while-loop it is! 🙂

Reply:
I've become a huge fan of `for { const x of items } { }`

Reply:
👍